macros/src/test/scala/code/chymyst/jc/GuardsSpec.scala

```diff
@@ -0,0 +1,52 @@
+package code.chymyst.jc
+
+import Macros.{m, b, go}
+import org.scalatest.{FlatSpec, Matchers}
+
+class GuardsSpec extends FlatSpec with Matchers {
+
+  behavior of "guard conditions"
+
+  it should "correctly recognize a guard condition with captured variables" in {
+    val a = m[Int]
+
+    val n = 10
+
+    val result = go { case a(x) if x > n => }
+
+    result.info.hasGuard shouldEqual GuardPresent(List(List('x))) // `n` should not be among the guard variables
+  }
+
+  it should "correctly recognize a guard condition with no variables" in {
+    val a = m[Int]
+
+    val n = 10
+
+    val result = go { case a(x) if 1 > n => }
+
+    result.info.hasGuard shouldEqual GuardPresent(List(List())) // `n` should not be among the guard variables
+  }
+
+  it should "correctly split a guard condition with three clauses" in {
```

Comment:
I did try that in the past (early saddle points) and it didn't work! Good stuff. If you need to have your own parser for guards, you may look at Li Haoyi's fastparse github (now maintained by Lars Hupel). If speed does not matter, I hear Scala lib combinator parsers are nice to use, not sure if they are good at error reporting (see discussion by Li Haoyi).